---
title: "Intro to MAF: Part I"
author: Nicole Black, Wade Boohar
date: 07/17/22
updated: 02/22/23
---

###### (1) Set Up

```{r setup}
    knitr::opts_knit$set(root.dir = normalizePath("/Users/kojiro/Desktop/qbio490/qbio_490_KainingFeng/analysis_data/")) 
```

```{r}
# install and load maftools, TCGAbiolinks, and ggplot2
if(!require(maftools)){
  install.packages("maftools")
}
library(maftools)
if(!require(TCGAbiolinks)){
  install.packages("TCGAbiolinks")
}
library(TCGAbiolinks)
if(!require(ggplot2)){
  install.packages("ggplot2")
}
library(ggplot2)
```

Before we get into MAF, we need to read in the clinical data (created in week 6) so it can be merged with our MAF data later on.

```{r Exercise 1.1}
clinical <- read.csv("/Users/kojiro/Desktop/qbio490/qbio_490_KainingFeng/analysis_data/brca_clinical_data.csv")
```

We also need to change the bcr_patient_barcode column name that way the MAF package can read our clinical file.

```{r Exercise 1.2}
colnames(clinical)[ colnames(clinical) == "bcr_patient_barcode" ] <- "Tumor_Sample_Barcode"
```

After changing the column name, re-save the clinical data frame to your computer.

```{r}
# change the file path! make sure it's in your week4 folder
write.csv(clinical, "/Users/kojiro/Desktop/qbio490/sp23_course_material/week7_maf/brca_clinical_data.csv", row.names = FALSE)
```

###### (2) MAF Package

Now, we can query in the MAF files using GDCquery()/GDCdownload()/GDCprepare() and read.maf(). 

```{r Exercise 2.1}
maf_query <- GDCquery(
    project = "TCGA-BRCA", 
    data.category = "Simple Nucleotide Variation", 
    access = "open", # we only have access to somatic mutations which are open access
    data.type = "Masked Somatic Mutation", 
    workflow.type = "Aliquot Ensemble Somatic Variant Merging and Masking"
)

GDCdownload(maf_query)

maf <- GDCprepare(maf_query) # as long as it runs, ignore any errors

maf_object <- read.maf(maf = maf, 
                       clinicalData = clinical,
                       isTCGA = TRUE)
```

###### (3) Exploring the MAF Object

Call str() on maf_object.

```{r Exercise 3.1}
str(maf_object)
```

Notice the @ and $ signs. What do each of them represent?

  @ represents:
  $ represents:
  
List all of the data frames within the maf_object by typing maf_object@ then either waiting for a popup, or if no popup appears, by hitting tab.

```{r Exercise 3.2}
# write code here
maf_object@data
```

Call head() and str() on the @data df and the @clinical.data df. These will be the main data frames we use in our MAF analysis.

```{r Exercise 3.3}
# write code here
str(maf_object@data)
head(maf_object@data)
```

View the first 110 items in Tumor_Sample_Barcode from @data and @clinical.data.

```{r Exercise 3.4}
# write code here
maf_object@data$Tumor_Sample_Barcode[1:110]
```

What do you notice about these two lists?

View the Tumor_Sample_Barcode column and the HGVSp_Short column for all TP53 mutations (i.e. Hugo_Symbol == "TP53"). Hint: boolean masking!

```{r Exercise 3.5}
# write code here
TP53_mask <- ifelse(maf_object@data$Hugo_Symbol =='TP53',T,F)

```

Call head() on the Amino_acids column in @data.

```{r Exercise 3.6}
# write code here
Head(maf_object@data$Amino_acids)
```
Refer to a codon chart and amino acid abbreviation chart to answer the following:
 - What mutation could cause a S/R change?
 Changing the last codon of Arg from a U/C to a A/G
 - What about an A/T change?
  Changing the middle codon of Ala from a C to a G

Call head() on the HGVSp_Short column in @data.

```{r Exercise 3.7}
# write code here
head(maf_object@data$HGVSp_Short)
```

What does p.P728Afs*11 mean? (Do some google-ing!)
P728 replaced by A and a frameshift is experienced; the new stop codon is 
termination 11.


###### (4) Fisher's Exact Test - Demo

Rewrite the vital_status and breast_carcinoma_estrogen_receptor_status columns of your clinical df as factors.

```{r Exercise 4.1}
clinical$vital_status <- factor(clinical$vital_status)

clinical$breast_carcinoma_estrogen_receptor_status <- factor(clinical$breast_carcinoma_estrogen_receptor_status,level=c('Positive','Negative'))
```

Create a contingency table using table().

```{r Exercise 4.2}
contig <- table(clinical$vital_status,clinical$breast_carcinoma_estrogen_receptor_status)
```

Visualize our contingency table using a mosaic plot.

```{r Exercise 4.3}
mosaicplot(contig)
```

Run our Fisher's exact test.

```{r Exercise 4.4}
fisher_test <- fisher.test(contig)
```

View the results of our test.

```{r Exercise 4.5}
fisher_test

fisher_test$p.value
```

What does this p-value tell us?
The two variables have significant dependence

Does this make sense given the biology of cancer?
This makes sense in biology of cancer because the more estrogen received, the more severe would the breaster cancer be. 

###### (6) Mutational Co-Occurence Example

Subset your maf_object to create geneA_maf and geneB_maf.

```{r Exercise 6.1}
# Choose either: TP53 + PIK3CA or MUC16 + RYR2

geneA_maf <- subsetMaf(maf = maf_object,
                       genes = 'MUC16') ## fill in with your gene name as a string

geneB_maf <- subsetMaf(maf = maf_object,
                       genes = 'RYR2')
```

Now let's start to build our contingency table.

Access the barcodes of the patients with mutations in genes A and the barcodes of the patients with mutations in gene B. How do we access patient barcodes within the MAF object without duplicate patients? @data or @clinical.data?

```{r Exercise 6.2}
mut_pats_geneA <- geneA_maf@clinical.data$Tumor_Sample_Barcode
mut_pats_geneB <- geneB_maf@clinical.data$Tumor_Sample_Barcode
```

Get the lengths of these two vectors.

```{r Exercise 6.3}
num_pats_geneA <- length(mut_pats_geneA)
num_pats_geneB <- length(mut_pats_geneB)
```

Use intersect() to get the patients with mutations in both gene A AND gene B. Then find the length of that vector. What does this tell us in our contingency table?
The probability of having mutation with both A and B is much lower than having a single mutation in one of the genes. 
```{r Exercise 6.4}
mut_pats_geneAB <- intersect(mut_pats_geneA,mut_pats_geneB)
num_AB <- length(mut_pats_geneAB)
num_AB
```

Now, determine the number of patients with one mutation and not the other. HINT: Do some subtraction with your above vectors.

```{r Exercise 6.5}
num_A_only <- num_pats_geneA - num_AB
num_B_only <- num_pats_geneB - num_AB
```

Finally, find the number of patients without either mutation. HINT: We know the total number of patients (based on our original maf_object), and we just calculated the number of patients with mutations (A only + B only + AB).

```{r Exercise 6.6}
num_neither <- length(maf_object@clinical.data$Tumor_Sample_Barcode)-num_A_only - num_B_only -num_AB
```

Create your contingency table:

```{r Exercise 6.7}
contig <- matrix(c(num_AB, 
                         num_B_only,
                         num_A_only,
                         num_neither), 
                       nrow=2)

# view the contingency table
contig

# view the contingency table as a mosaic
mosaicplot(contig)
```

Run your Fisher exact test!

```{r Exercise 6.8}
fisher_test <- fisher.test(contig)
fisher_test
```

Are your mutations co-occurent or mutually exclusive?
They are very likely to be mutually exclusive. 
###### (7) Maftools somaticInteractions()

Luckily, we don't have to do the above work by hand for each gene pair we want to look at because maftools has a function that does it for us!

Run somaticInteractions for your above genes by passing in a vector with your gene names into the genes argument.

```{r Exercise 7.1}
somaticInteractions(maf = maf_object,
                    genes = c('MUC16','RYR2'),
                    pvalue = c(0.05, 0.1)) ## these are the default p-value significance levels
```

Run the following line of code for the top 25 genes.

```{r Exercise 7.2}
somaticInteractions(maf = maf_object,
                    top = 25,
                    pvalue = c(0.05, 0.1))
```


###### (8) Oncoplots

We are going to start off by making our first MAF figure using the oncoplot() function. Each column on an oncoplot represents an individual patient, and each row represents a gene of interest. The different colors represent the different types of mutations where grey represents no mutation in that gene for that patient.

Generate an oncoplot by filling in the maf argument below. Fill in the top argument with various integer arguments (e.g. 5, 10, 20, etc.) to see the genes that are mutated in the most patients. Save your plot to your week7_maf folder.

```{r Exercise 8.1}
oncoplot(maf = maf_object,
         top = 10,
         borderCol = NA) 

ggsave("/Users/kojiro/Desktop/qbio490/sp23_course_material/week7_maf/filename.png")
```


You can also create an oncoplot for specific genes of your choosing by passing a vector into oncoplot genes argument. Pick 5 genes to plot. Hint: the genes argument accepts a vector of strings.

```{r Exercise 8.2}
oncoplot(maf = maf_object,
         genes = c('MUC16','RYR2', 'TP53','FLG', 'PIK3CA'),
         borderCol = NA)
```


We can further customize our oncoplots by adding clinical data in the form of an annotation. Let's practice this by adding age_category to our oncoplot. For ease, the code to create the 
age_category column has been added but if you don't understand it, review the Intro_to_R notebook.

```{r Exercise 8.3}
median_age <- median(as.numeric(maf_object@clinical.data$age_at_initial_pathologic_diagnosis))
print(median_age)

maf_object@clinical.data$age_category <- ifelse(maf_object@clinical.data$age_at_initial_pathologic_diagnosis > median_age,'Old', 'Young')

oncoplot(maf = maf_object,
         top = 10,
         clinicalFeatures = "age_category",
         borderCol = NA)
```

Do you see any clustering between old and young patients?
No, they seem perfectly mixed
###### (9) subsetMaf + CoOncoplots

Another way to compare between patient populations is to use a coOncoplot. Before we create our coOncoplot, we first need to subset our maf_object dataframe into two separate data frames, one for old patients and one for young patients.

We can do this with subsetMaf(). The subsetMaf(maf, tsb) function needs two arguments in this case: maf is the original MAF object, and tsb is a vector of patient IDs (barcodes) to select.

Store the young patient barcodes in a vector called young_patient_barcodes. Hint: you can use boolean masking for this!

```{r Exercise 9.1}
mask <-ifelse(maf_object@clinical.data$age_catagory == 'Young',T,F)
young_patient_barcodes <- maf_object@clinical.data$Tumor_Sample_Barcode[mask]
```

Use the young_patient_barcode vector in the tsb argument of subsetMaf.

```{r Exercise 9.2}
young_maf <- subsetMaf(maf = maf_object,
                       tsb = young_patient_barcodes)
```

Create an old_patient_barcode vector and use that to subset just the old patient data.

```{r Exercise 9.3}
old_mask <- ifelse(maf_object@clinical.data$age_category == 'Old',T,F)
old_patient_barcodes <- maf_object@clinical.data$Tumor_Sample_Barcode[old_mask]
  
old_maf <- subsetMaf(maf = maf_object,
                       tsb = old_patient_barcodes)
```

Use coOncoplot() to draw two oncoplots side by side, comparing young vs. old patients. m1 is the first MAF object to plot, and m2 is the second. Be sure to provide descriptive names in the m1Name and m2Name arguments! Again, save the plot using the ggsave() function.

```{r Exercise 9.4}
coOncoplot(m1 = young_maf, 
           m2 = old_maf, 
           m1Name = 'Young patients <= 58', 
           m2Name = 'Old patients > 58', 
           borderCol = NA)

ggsave("/Users/kojiro/Desktop/qbio490/sp23_course_material/week7_maf/filename.png")
```

Are there any genes that seem to be mutated at different rates between the two populations? If so, pick one. Can you think of any reasons why that gene might be mutated more in younger/older patients (Google it!)

Chosen Gene:TP53
More highly mutated in young or old population: young population
Hypothesized Reason: The TP53 is an important gene, which may shorten the lifetime if it mutates. Thus, the young with a mutation in TP53 will die before they become a part of the old population.

###### (10) Lollipop Plots

Lollipop plots display the location of each mutation along the gene’s DNA sequence. The x-axis of a lollipop plot is the amino acid sequence that makes up the gene. Individual mutations are annotated by a ball-and-stick, which looks like a lollipop, hence the name. The colors represents the type of mutation at a given point on the gene (similar to colors of the oncoplots), and the visualization in the middle of the plot is the gene itself differentiated between its different domain/motif sections.

Let's make a lollipop plot for TP53.

```{r Exercise 10.1}
lollipopPlot(maf = maf_object,
             gene = "MUC16")

ggsave("/Users/kojiro/Desktop/qbio490/sp23_course_material/week7_maf/filename.png")
```


Similarly to how we created co-Oncoplots to compare the oncoplot data between two different populations, we can also create co-lollipop plots to do the same. We will use our subsetted
maf objects from before.

Now fill in the co-lollipop plot function below. Be sure to add descriptive names for each population.

```{r Exercise 10.3}
lollipopPlot2(m1 = young_maf, 
              m2 = old_maf, 
              m1_name = 'Young patients <= 58',
              m2_name = 'Old patients > 58',
              gene = "MUC16") ## pick any gene of your choosing to fill in here
```

Now, interpret your plot. Answer the following questions:

- Does it look like the gene is more commonly mutated in young or old patients?
A slightly more commonly mutated in old patients. 
- Are there more mutations in a specific region of the protein? Why might that be?
Yes, in some regions the mutation are condensed, while in other places are sparse. This might be the codons in some regions are easily being replaced by other codons which can lead to a change in amino acid. 
- What kinds of mutations are the most common?
Missense_Mutation
- Are there any other interesting observations?
There's a point of obvious  missense mutation whihc is hgher than the other mutations in both old and young patients. 

###### (11) Survival Analysis (mafSurvival)

Remember back to Kaplan-Meier survival plots (from the clinical data tutorial). These plots allow us to visualize survival probabilities stratified by different clinical factors (such as age, sex, stage, etc.)

We can also visualize survival probabilities stratified by mutational status. To do this, we use mafSurvival. This is very similar to our traditional KM plots, you may want to open up the clinical data tutorial to follow along there as well.

```{r Exercise 11.1}
# adding death event
maf_object@clinical.data$Overall_Survival_Status <-
  ifelse(maf_object@clinical.data$vital_status == "Alive",
         maf_object@clinical.data$Overall_Survival_Status <- TRUE,
         maf_object@clinical.data$Overall_Survival_Status <- FALSE)
mafSurvival(maf = maf_object,
            genes = "KMT2C", ## pick a gene of your choosing
            time = "days_to_last_followup", ## name of the column in maf_object@clinical.data containing survival time
            Status = "Overall_Survival_Status", ## name of the column that contains a boolean value for death events, you may need to recreate this... 
            isTCGA = TRUE)

```

Provide a brief description of your chosen gene and it's posited relationship to survival.
The plot shows the overall survivability of cancer in patients with wild-type
KMT2C vs. mutant KMT2C. It seems like mutant KMT2C lowers the survivability of
cancer.

What can you conclude from your plot?
Using the conventional p-value cutoff of 0.05, there is not a statiscally significant relationship between KMT2C mutation status and cancer survivability.

Is this what you would expect according to the literature? Why or why not?
According to literature, mutated KMT2C is implicated with "shorter progression-free survival on anti-estrogen therapy". KMT2C appears to promote cancer proliferation.
This trend can be observed, although not statistically significant, with the data on hand. With a larger sample size, this trend would likely become statiscally significant.
See: Gala et al. "KMT2C mediates the estrogen dependence of breast cancer through regulation of ERα enhancer function"
https://www.nature.com/articles/s41388-018-0273-5

